File: backend/name_gen.py
"""
Anonymous name generation: business description -> brandable name candidates ->
live availability check, gated by trial tier (curious/trying/engaged).

Identity is multi-signal: visitor_id (client cookie), fingerprint (FingerprintJS
visitorId), and salted IP hash. Usage is aggregated by MAX across all three so
that defeating any single signal does not reset the counter.

Generation uses Anthropic Claude when ANTHROPIC_API_KEY is set, otherwise falls
back to a deterministic rule-based generator so the endpoint never hard-fails.
"""

import asyncio
import hashlib
import json
import logging
import os
import re
import secrets
import time
import uuid
from datetime import datetime, timezone
from typing import Optional

from queries import get_db_conn
from valkey_client import create_split_job, get_job_status, get_job_results

logger = logging.getLogger(__name__)

# ── Trial tier limits ──────────────────────────────────────────────────────
CURIOUS_LIMIT = 5
TRYING_LIMIT = 10
FULL_RESULT_COUNT = 36
ENGAGED_VISIBLE_COUNT = 3
ENGAGED_COOLDOWN_MS = 30_000
GLOBAL_COOLDOWN_MS = 5_000   # min spacing between any two generations
ROLLING_WINDOW_DAYS = 7

# BYOK callers don't burn our LLM budget — but they still hit the availability
# engine, so we cap them separately, more permissively. Tunable.
BYOK_DAILY_LIMIT = 50

# Hosted-LLM anon caps (cost-protection layer in front of the home machine).
# Per-visitor cap drives signup conversion; per-IP cap defends against scripted
# abuse that rotates visitor_ids. Adjust via env without redeploying code.
HOSTED_VISITOR_DAILY_LIMIT = int(os.environ.get('HOME_LLM_VISITOR_DAILY', '5'))
HOSTED_IP_DAILY_LIMIT = int(os.environ.get('HOME_LLM_IP_DAILY', '20'))

# Anthropic config (optional, legacy fallback before HOME_LLM)
ANTHROPIC_API_KEY = os.environ.get('ANTHROPIC_API_KEY', '')
ANTHROPIC_MODEL = os.environ.get('NAMEGEN_MODEL', 'claude-haiku-4-5')

TLD_BUCKETS: dict[str, list[str]] = {
    'com_only': ['com'],
    'tech':     ['io', 'dev', 'ai', 'app'],
    'global':   ['co', 'net', 'org', 'com'],
    'any':      ['com', 'io', 'co', 'ai', 'dev', 'app', 'xyz'],
}

PORTAL_URL = os.environ.get('PORTAL_URL', 'https://portal.canyougrab.it')
SIGNUP_URL = f'{PORTAL_URL}/signup'


# ── IP hashing with daily salt rotation ────────────────────────────────────

def _today() -> str:
    return datetime.now(timezone.utc).strftime('%Y-%m-%d')


def get_or_create_daily_salt() -> str:
    today = _today()
    conn = get_db_conn()
    try:
        with conn.cursor() as cur:
            cur.execute('SELECT salt FROM anon_ip_salt WHERE salt_date = %s', (today,))
            row = cur.fetchone()
            if row:
                return row[0]
            new_salt = secrets.token_hex(32)
            cur.execute(
                'INSERT INTO anon_ip_salt (salt_date, salt) VALUES (%s, %s) '
                'ON CONFLICT (salt_date) DO NOTHING RETURNING salt',
                (today, new_salt),
            )
            ret = cur.fetchone()
            conn.commit()
            if ret:
                return ret[0]
            cur.execute('SELECT salt FROM anon_ip_salt WHERE salt_date = %s', (today,))
            return cur.fetchone()[0]
    finally:
        conn.close()


def hash_ip(ip: str) -> Optional[str]:
    if not ip:
        return None
    salt = get_or_create_daily_salt()
    return hashlib.sha256(f'{salt}:{ip}'.encode()).hexdigest()


# ── Visitor identity & tier aggregation ────────────────────────────────────

def aggregate_usage(visitor_id: str, fingerprint: Optional[str], ip_hash: Optional[str]) -> dict:
    """Return {count, last_at} aggregated across (visitor_id OR fingerprint OR ip_hash)
    within the rolling window. Count is the highest of the three lookups."""
    conn = get_db_conn()
    try:
        with conn.cursor() as cur:
            counts = []
            last_at = None

            def _q(field: str, value: str):
                cur.execute(
                    f'SELECT COUNT(*), MAX(created_at) FROM anon_name_gen_usage '
                    f"WHERE {field} = %s AND created_at > NOW() - INTERVAL '%s days'",
                    (value, ROLLING_WINDOW_DAYS),
                )
                return cur.fetchone()

            for field, value in (
                ('visitor_id', visitor_id),
                ('fingerprint', fingerprint),
                ('ip_hash', ip_hash),
            ):
                if not value:
                    continue
                c, m = _q(field, value)
                counts.append(c or 0)
                if m and (last_at is None or m > last_at):
                    last_at = m

            return {'count': max(counts) if counts else 0, 'last_at': last_at}
    finally:
        conn.close()


def record_usage(visitor_id: str, fingerprint: Optional[str], ip_hash: Optional[str]) -> None:
    conn = get_db_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                'INSERT INTO anon_name_gen_usage (visitor_id, fingerprint, ip_hash) '
                'VALUES (%s, %s, %s)',
                (visitor_id, fingerprint, ip_hash),
            )
            conn.commit()
    finally:
        conn.close()


def daily_count_visitor(visitor_id: str, fingerprint: Optional[str]) -> int:
    """Generations from this visitor in the last 24h. Aggregates by MAX across
    visitor_id and fingerprint so clearing cookies alone doesn't reset."""
    conn = get_db_conn()
    try:
        with conn.cursor() as cur:
            counts: list[int] = []
            for field, value in (('visitor_id', visitor_id), ('fingerprint', fingerprint)):
                if not value:
                    continue
                cur.execute(
                    f'SELECT COUNT(*) FROM anon_name_gen_usage '
                    f"WHERE {field} = %s AND created_at > NOW() - INTERVAL '24 hours'",
                    (value,),
                )
                counts.append(cur.fetchone()[0] or 0)
            return max(counts) if counts else 0
    finally:
        conn.close()


def daily_count_ip(ip_hash: Optional[str]) -> int:
    """Generations from this IP-hash in the last 24h. Used to gate scripted
    abuse that rotates visitor_ids."""
    if not ip_hash:
        return 0
    conn = get_db_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                'SELECT COUNT(*) FROM anon_name_gen_usage '
                "WHERE ip_hash = %s AND created_at > NOW() - INTERVAL '24 hours'",
                (ip_hash,),
            )
            return cur.fetchone()[0] or 0
    finally:
        conn.close()


# ── Saved generation lists (claim-on-signup) ───────────────────────────────

def save_generation_list(visitor_id: str, description: str, payload: dict) -> Optional[str]:
    """Persist a generation result so it can be attached to an account when
    the visitor signs up. Returns the new list id, or None on failure."""
    if not visitor_id or not description:
        return None
    conn = get_db_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                'INSERT INTO name_generation_lists (visitor_id, description, payload) '
                'VALUES (%s, %s, %s) RETURNING id',
                (visitor_id, description, json.dumps(payload)),
            )
            row = cur.fetchone()
            conn.commit()
            return str(row[0]) if row else None
    except Exception as e:
        logger.warning('Failed to save generation list: %s', e)
        conn.rollback()
        return None
    finally:
        conn.close()


def claim_anon_lists(visitor_id: str, user_sub: str) -> int:
    """Attach all unclaimed anon lists for `visitor_id` to `user_sub`.
    Returns count of rows claimed. Idempotent: only touches NULL user_sub rows.
    """
    if not visitor_id or not user_sub:
        return 0
    conn = get_db_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                'UPDATE name_generation_lists '
                'SET user_sub = %s, claimed_at = NOW() '
                'WHERE visitor_id = %s AND user_sub IS NULL',
                (user_sub, visitor_id),
            )
            n = cur.rowcount
            conn.commit()
            if n:
                logger.info('Claimed %d anon lists for user_sub=%s visitor=%s', n, user_sub, visitor_id[:8])
            return n
    except Exception as e:
        logger.error('Failed to claim anon lists: %s', e)
        conn.rollback()
        return 0
    finally:
        conn.close()


def get_saved_list(share_id: str) -> Optional[dict]:
    """Public read of a saved generation list. No auth — these are shareable.
    Contents are public-safe (suggested names + availability), no PII.
    Strips per-viewer fields like `locked` so shared viewers see everything."""
    if not share_id:
        return None
    conn = get_db_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                'SELECT description, payload, created_at '
                'FROM name_generation_lists WHERE id::text = %s',
                (share_id,),
            )
            row = cur.fetchone()
            if not row:
                return None
            payload = row[1] or {}
            results = payload.get('results') or []
            for r in results:
                r.pop('locked', None)
            payload['results'] = results
            return {
                'description': row[0],
                'payload': payload,
                'created_at': row[2].isoformat() if row[2] else None,
            }
    finally:
        conn.close()


def list_user_generations(user_sub: str, limit: int = 50) -> list[dict]:
    """Return saved lists owned by `user_sub`, newest first."""
    if not user_sub:
        return []
    conn = get_db_conn()
    try:
        with conn.cursor() as cur:
            cur.execute(
                'SELECT id, description, payload, created_at, claimed_at '
                'FROM name_generation_lists '
                'WHERE user_sub = %s '
                'ORDER BY created_at DESC LIMIT %s',
                (user_sub, limit),
            )
            return [
                {
                    'id': str(r[0]),
                    'description': r[1],
                    'payload': r[2],
                    'created_at': r[3].isoformat() if r[3] else None,
                    'claimed_at': r[4].isoformat() if r[4] else None,
                }
                for r in cur.fetchall()
            ]
    finally:
        conn.close()


def tier_for_count(count: int) -> str:
    if count <= CURIOUS_LIMIT:
        return 'curious'
    if count <= TRYING_LIMIT:
        return 'trying'
    return 'engaged'


def cooldown_remaining_ms(tier: str, last_at: Optional[datetime]) -> int:
    if last_at is None:
        return 0
    elapsed_ms = (datetime.now(timezone.utc) - last_at).total_seconds() * 1000
    floor = ENGAGED_COOLDOWN_MS if tier == 'engaged' else GLOBAL_COOLDOWN_MS
    return max(0, int(floor - elapsed_ms))


# ── Brandable name generation ──────────────────────────────────────────────

LLM_PROMPT = """You are a brand-naming expert. Generate {count} candidate brand names for this business:

DESCRIPTION:
{description}

STYLES: {styles}
PREFERRED EXTENSION TYPE: {tld_pref}

Rules:
- Each name is a single lowercase word or two words concatenated, alphanumeric only
- 4-18 characters
- Avoid generic words ("app", "platform", "startup", "company")
- Mix invented words, evocative metaphors, compound words, and direct descriptors
- Avoid offensive or trademark-likely names

Return ONLY a JSON array of strings, no commentary. Example: ["frondly", "treekit", "leafgraph"]"""


async def llm_generate_bases_async(description: str, styles: list[str], tld_pref: str, count: int) -> list[str]:
    """Async path: prefer the hosted home LLM, fall back to Anthropic, then rule-based.

    The hosted LLM is the primary anon-mode generator. Anthropic stays wired as a
    legacy fallback for environments that have ANTHROPIC_API_KEY set but not
    HOME_LLM_API_KEY. Both fall back to rule-based on any failure so the
    endpoint never hard-fails for the visitor.
    """
    # Hosted home-LLM path
    try:
        from hosted_llm import generate_bases as hosted_generate, is_configured
        if is_configured():
            bases = await hosted_generate(description, styles, tld_pref, count)
            if bases:
                return bases
            logger.info('Hosted LLM returned no bases; falling back')
    except Exception as e:
        # HostedQueueFullError, HostedUnavailableError, network, etc.
        logger.warning('Hosted LLM unavailable (%s); falling back', type(e).__name__)

    # Anthropic legacy fallback
    if ANTHROPIC_API_KEY:
        try:
            return _anthropic_generate_bases(description, styles, tld_pref, count)
        except Exception as e:
            logger.warning('Anthropic fallback failed (%s); using rules', e)

    logger.info('No LLM available; using rule-based name generator')
    return rule_based_bases(description, styles, count)


def _anthropic_generate_bases(description: str, styles: list[str], tld_pref: str, count: int) -> list[str]:
    import anthropic  # type: ignore
    client = anthropic.Anthropic(api_key=ANTHROPIC_API_KEY)
    msg = client.messages.create(
        model=ANTHROPIC_MODEL,
        max_tokens=1024,
        messages=[{
            'role': 'user',
            'content': LLM_PROMPT.format(
                count=count,
                description=description.strip()[:1000],
                styles=', '.join(styles) if styles else 'modern',
                tld_pref=tld_pref,
            ),
        }],
    )
    text = msg.content[0].text.strip()
    text = re.sub(r'^```(?:json)?\s*|\s*```$', '', text, flags=re.MULTILINE).strip()
    bases = json.loads(text)
    if not isinstance(bases, list):
        raise ValueError('LLM did not return a list')
    return [_clean(b) for b in bases if isinstance(b, str)][:count]


_STOPWORDS = frozenset({
    'a', 'an', 'and', 'the', 'for', 'to', 'of', 'in', 'on', 'with', 'that',
    'this', 'is', 'it', 'we', 'our', 'i', 'my', 'be', 'by', 'as', 'at', 'or',
    'from', 'but', 'are', 'was', 'were', 'will', 'would', 'can', 'could',
    'should', 'have', 'has', 'had', 'you', 'your', 'they', 'their', 'them',
    'app', 'platform', 'service', 'company', 'business', 'startup',
})


def _clean(s: str) -> str:
    return re.sub(r'[^a-z0-9]', '', s.lower())[:20]


def rule_based_bases(description: str, styles: list[str], count: int) -> list[str]:
    tokens = [
        t for t in re.findall(r'[a-z]+', description.lower())
        if len(t) >= 3 and t not in _STOPWORDS
    ][:8]
    bases: list[str] = []
    seen: set[str] = set()

    def add(b: str):
        b = _clean(b)
        if 3 <= len(b) <= 20 and b not in seen:
            seen.add(b)
            bases.append(b)

    for t in tokens:
        add(t)

    if 'short' in styles:
        for t in tokens:
            add(t[: max(4, int(len(t) * 0.7))])

    if 'playful' in styles:
        for t in tokens:
            add(f'{t}ly')
            add(f'get{t}')

    if 'wordplay' in styles:
        for t in tokens:
            add(t.replace('s', 'z') if t.endswith('s') else f'{t}ify')

    if 'compound' in styles or len(tokens) >= 2:
        for i, a in enumerate(tokens):
            for j, b in enumerate(tokens):
                if i != j:
                    add(f'{a}{b}')

    for t in tokens[:3]:
        for p in ('use', 'try', 'join', 'go'):
            add(f'{p}{t}')
        for s in ('hq', 'labs', 'kit', 'hub'):
            add(f'{t}{s}')

    return bases[:count]


def expand_to_domains(bases: list[str], tld_pref: str, cap: int) -> list[str]:
    tlds = TLD_BUCKETS.get(tld_pref, TLD_BUCKETS['any'])
    out: list[str] = []
    for base in bases:
        for tld in tlds:
            if len(out) >= cap:
                return out
            out.append(f'{base}.{tld}')
        if len(out) >= cap:
            return out
    return out


# ── Domain availability via existing job pipeline ──────────────────────────

POLL_INTERVAL = 0.3
POLL_TIMEOUT = 30.0


async def check_domains_anon(domains: list[str]) -> list[dict]:
    """Run domains through the same job pipeline as /api/check/bulk, but
    without quota tracking. Used only for trial-gated anonymous traffic.
    """
    job_id = str(uuid.uuid4())
    consumer = f'anon:{job_id[:8]}'
    create_split_job(job_id, consumer, domains)

    deadline = time.monotonic() + POLL_TIMEOUT
    while time.monotonic() < deadline:
        await asyncio.sleep(POLL_INTERVAL)
        job = get_job_status(job_id)
        if job is None:
            continue
        if job['status'] == 'completed':
            return get_job_results(job_id) or []
        if job['status'] == 'failed':
            logger.warning('Anon domain check failed: %s', job.get('error'))
            return []
    logger.warning('Anon domain check timed out for job %s', job_id[:8])
    return []


# ── Top-level pipeline ─────────────────────────────────────────────────────

async def generate_for_visitor(
    description: str,
    styles: list[str],
    tld_pref: str,
    visitor_id: str,
    fingerprint: Optional[str],
    ip_hash: Optional[str],
) -> dict:
    """Full pipeline. Returns the response dict, or raises a structured error
    on hard limits / cooldowns / hosted-LLM unavailability.
    """
    # Per-IP daily cap protects the hosted LLM from scripted abuse that rotates
    # visitor_ids — checked first, since IP is harder to spoof than a cookie.
    if daily_count_ip(ip_hash) >= HOSTED_IP_DAILY_LIMIT:
        raise HostedDailyCapError('ip', HOSTED_IP_DAILY_LIMIT)
    # Per-visitor daily cap drives signup conversion.
    if daily_count_visitor(visitor_id, fingerprint) >= HOSTED_VISITOR_DAILY_LIMIT:
        raise HostedDailyCapError('visitor', HOSTED_VISITOR_DAILY_LIMIT)

    pre = aggregate_usage(visitor_id, fingerprint, ip_hash)
    pre_tier = tier_for_count(pre['count'])
    cooldown = cooldown_remaining_ms(pre_tier, pre['last_at'])
    if cooldown > 0:
        raise CooldownError(cooldown)

    record_usage(visitor_id, fingerprint, ip_hash)
    new_count = pre['count'] + 1
    tier = tier_for_count(new_count)

    bases = await llm_generate_bases_async(description, styles, tld_pref, count=18)
    domains = expand_to_domains(bases, tld_pref, cap=FULL_RESULT_COUNT)

    raw = await check_domains_anon(domains) if domains else []
    by_domain = {r['domain']: r for r in raw}

    results = []
    for d in domains:
        r = by_domain.get(d, {})
        base = d.rsplit('.', 1)[0]
        tld = d.rsplit('.', 1)[1] if '.' in d else ''
        results.append({
            'domain': d,
            'available': r.get('available'),
            'tld': tld,
            'base': base,
            'rationale': None,
        })

    # Sort: available first, shorter domains first
    results.sort(key=lambda r: (
        0 if r['available'] is True else (1 if r['available'] is None else 2),
        len(r['domain']),
    ))

    visible = ENGAGED_VISIBLE_COUNT if tier == 'engaged' else len(results)
    for i, r in enumerate(results):
        if i >= visible:
            r['locked'] = True

    limit = (
        CURIOUS_LIMIT if tier == 'curious'
        else TRYING_LIMIT if tier == 'trying'
        else TRYING_LIMIT
    )
    cooldown_ms = ENGAGED_COOLDOWN_MS if tier == 'engaged' else 0

    response = {
        'results': results,
        'description': description,
        'tier': tier,
        'generations_used': new_count,
        'generations_limit': limit,
        'signup_url': SIGNUP_URL,
        'cooldown_ms': cooldown_ms,
    }
    list_id = save_generation_list(
        visitor_id=visitor_id,
        description=description,
        payload={
            'styles': styles,
            'tld_pref': tld_pref,
            'results': results,
            'tier': tier,
        },
    )
    if list_id:
        response['list_id'] = list_id
    return response


class CooldownError(Exception):
    def __init__(self, retry_after_ms: int):
        self.retry_after_ms = retry_after_ms
        super().__init__(f'Cooldown active, retry in {retry_after_ms}ms')


class HostedDailyCapError(Exception):
    """Anonymous visitor has hit the hosted-LLM daily cap. The route surfaces
    this as a 429 with `signup_url` so the FE can render its soft paywall.

    `scope` is "visitor" or "ip" — the FE messages slightly differently."""
    def __init__(self, scope: str, daily_limit: int):
        self.scope = scope
        self.daily_limit = daily_limit
        super().__init__(f'Hosted LLM {scope} daily cap reached ({daily_limit})')

File: portal/src/shared/name-generator/nameGen.ts
import { API_BASE_URL, PORTAL_URL } from "@/config";
import { logger } from "@/lib/logger";
import { getVisitorHeaders } from "./visitor";
import {
  TIER_LIMITS,
  incrementUsage,
  readUsage,
  tierFor,
  type Tier,
  type UsageState,
} from "./usage";
import { readByokKey } from "./byok/storage";
import { getAdapter } from "./byok/providers";
import { withVisitorId } from "./visitor";

export type NameStyle = "modern" | "playful" | "professional" | "short" | "wordplay" | "compound";
export type TldPreference = "com_only" | "tech" | "global" | "any";

export interface GenerateNamesRequest {
  description: string;
  styles?: NameStyle[];
  tldPreference?: TldPreference;
  count?: number;
  /** Bases the user liked — biases the LLM toward similar style. */
  anchors?: string[];
  /** Bases already shown — the LLM is told not to repeat them. */
  exclude?: string[];
}

export interface GeneratedName {
  domain: string;
  available: boolean | null;
  rationale?: string;
  tld: string;
  base: string;
  locked?: boolean;
}

export type GenerationMode = "hosted" | "byok";

export interface GenerateNamesResponse {
  results: GeneratedName[];
  description: string;
  shareId?: string;
  listId?: string;
  tier: Tier;
  generationsUsed: number;
  generationsLimit: number;
  signupUrl: string;
  cooldownMs: number;
  mode: GenerationMode;
  providerLabel?: string;
}

export class TrialExhaustedError extends Error {
  signupUrl: string;
  retryAfterMs: number;
  constructor(signupUrl: string, retryAfterMs: number) {
    super("Free trial limit reached");
    this.signupUrl = signupUrl;
    this.retryAfterMs = retryAfterMs;
  }
}

/**
 * Server returned 429 from /api/names/check — the BYOK visitor has hit the
 * daily availability-check cap. Surfaces a graceful soft paywall in the UI.
 */
export class ByokLimitError extends Error {
  signupUrl: string;
  dailyLimit: number;
  constructor(signupUrl: string, dailyLimit: number) {
    super("BYOK daily limit reached");
    this.signupUrl = signupUrl;
    this.dailyLimit = dailyLimit;
  }
}

const TLD_BUCKETS: Record<TldPreference, string[]> = {
  com_only: ["com"],
  tech: ["io", "dev", "ai", "app"],
  global: ["co", "net", "org", "com"],
  any: ["com", "io", "co", "ai", "dev", "app", "xyz"],
};

const DEMO_KEY = import.meta.env.VITE_DEMO_API_KEY ?? "";
const NAMEGEN_PATH = "/api/names/generate";
const CHECK_PATH = "/api/names/check";

// Higher caps for BYOK callers — they aren't burning our LLM budget, only
// our availability engine. Tunable.
const BYOK_DAILY_SOFT_LIMIT = 50;
const BYOK_NUDGE_AT = 10;

export async function generateNames(req: GenerateNamesRequest): Promise<GenerateNamesResponse> {
  const byok = readByokKey();
  if (byok && byok.key) {
    return generateWithByok(req, byok);
  }

  // BYOK is required to generate. The UI gates submission, but throw a clear
  // error if anything calls this without a key as defense in depth.
  throw new Error("Add your AI key (Anthropic, OpenAI, or Google AI Studio) to generate names.");

  // Hosted-mode generation is currently disabled — kept commented for future
  // re-enablement when we add a paid hosted tier.
  // eslint-disable-next-line no-unreachable
  const visitorHeaders = await getVisitorHeaders();

  try {
    const res = await fetch(`${API_BASE_URL}${NAMEGEN_PATH}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...visitorHeaders,
        ...(DEMO_KEY ? { Authorization: `Bearer ${DEMO_KEY}` } : {}),
      },
      body: JSON.stringify({
        description: req.description,
        styles: req.styles ?? [],
        tld_preference: req.tldPreference ?? "any",
        count: req.count ?? TIER_LIMITS.fullResultCount,
      }),
    });

    if (res.status === 404 || res.status === 501) {
      logger.warn("Name generation endpoint not yet implemented; using local fallback");
      return mockGenerate(req);
    }

    if (res.status === 429) {
      const data = await res.json().catch(() => ({}));
      throw new TrialExhaustedError(
        withVisitorId(data?.signup_url ?? `${PORTAL_URL}/signup`),
        Number(data?.retry_after_ms ?? 0),
      );
    }

    if (!res.ok) {
      const data = await res.json().catch(() => ({}));
      const detail = data?.detail ?? data?.message ?? `HTTP ${res.status}`;
      throw new Error(typeof detail === "string" ? detail : JSON.stringify(detail));
    }

    const raw = (await res.json()) as GenerateNamesResponse & { list_id?: string };
    return { ...raw, listId: raw.listId ?? raw.list_id, mode: raw.mode ?? "hosted" };
  } catch (err) {
    if (err instanceof TrialExhaustedError) throw err;
    if (err instanceof TypeError) {
      logger.warn("Name generation network error; falling back to mock", { error: String(err) });
      return mockGenerate(req);
    }
    throw err;
  }
}

async function generateWithByok(
  req: GenerateNamesRequest,
  byok: { provider: import("./byok/types").ProviderId; key: string; model: string },
): Promise<GenerateNamesResponse> {
  const adapter = getAdapter(byok.provider);
  const tldPref = req.tldPreference ?? "any";
  const styles = req.styles ?? [];
  const fullCount = req.count ?? TIER_LIMITS.fullResultCount;

  let bases: string[] = [];
  try {
    bases = await adapter.generateBases({
      key: byok.key,
      model: byok.model || adapter.meta.defaultModel,
      description: req.description,
      styles: styles.map(String),
      tldPref,
      count: 18,
      anchors: req.anchors,
      exclude: req.exclude,
    });
  } catch (err) {
    const msg = err instanceof Error ? err.message : "Provider call failed";
    throw new Error(`Your ${adapter.meta.label} key didn't work: ${msg}`);
  }

  const tlds = TLD_BUCKETS[tldPref];
  const domains: string[] = [];
  for (const base of bases) {
    for (const tld of tlds) {
      if (domains.length >= fullCount) break;
      domains.push(`${base}.${tld}`);
    }
    if (domains.length >= fullCount) break;
  }

  const checked = await fetchAvailability(domains);
  const byDomain = new Map(checked.map((r) => [r.domain, r] as const));

  const results: GeneratedName[] = domains.map((d) => {
    const r = byDomain.get(d);
    const dot = d.lastIndexOf(".");
    return {
      domain: d,
      available: r?.available ?? null,
      tld: dot >= 0 ? d.slice(dot + 1) : "",
      base: dot >= 0 ? d.slice(0, dot) : d,
    };
  });

  results.sort((a, b) => {
    const score = (r: GeneratedName) =>
      r.available === true ? 0 : r.available === null ? 1 : 2;
    const s = score(a) - score(b);
    return s !== 0 ? s : a.domain.length - b.domain.length;
  });

  // BYOK uses a separate counter purely for soft-nudge timing; lives in localStorage.
  const used = bumpByokCounter();

  return {
    results,
    description: req.description,
    tier: "curious",
    generationsUsed: used,
    generationsLimit: BYOK_DAILY_SOFT_LIMIT,
    signupUrl: withVisitorId(`${PORTAL_URL}/signup`),
    cooldownMs: 0,
    mode: "byok",
    providerLabel: adapter.meta.label,
  };
}

interface CheckResultRow {
  domain: string;
  available: boolean | null;
}

async function fetchAvailability(domains: string[]): Promise<CheckResultRow[]> {
  if (domains.length === 0) return [];
  const visitorHeaders = await getVisitorHeaders();
  try {
    const res = await fetch(`${API_BASE_URL}${CHECK_PATH}`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...visitorHeaders,
      },
      body: JSON.stringify({ domains, mode: "byok" }),
    });
    if (res.status === 404 || res.status === 501) {
      logger.warn("Availability-only endpoint not yet implemented; faking results");
      return domains.map((d) => ({ domain: d, available: pseudoAvailable(d) }));
    }
    if (res.status === 429) {
      const data = await res.json().catch(() => ({}));
      throw new ByokLimitError(
        data?.signup_url ?? `${PORTAL_URL}/signup`,
        Number(data?.daily_limit ?? BYOK_DAILY_SOFT_LIMIT),
      );
    }
    if (!res.ok) {
      const data = await res.json().catch(() => ({}));
      const detail = data?.detail ?? data?.message ?? `HTTP ${res.status}`;
      throw new Error(typeof detail === "string" ? detail : JSON.stringify(detail));
    }
    const data = await res.json();
    return data.results as CheckResultRow[];
  } catch (err) {
    if (err instanceof ByokLimitError) throw err;
    if (err instanceof TypeError) {
      logger.warn("Availability check network error; faking results", { error: String(err) });
      return domains.map((d) => ({ domain: d, available: pseudoAvailable(d) }));
    }
    throw err;
  }
}

const BYOK_COUNTER_KEY = "cygi_byok_count_v1";
function bumpByokCounter(): number {
  try {
    const raw = localStorage.getItem(BYOK_COUNTER_KEY);
    const today = new Date().toISOString().slice(0, 10);
    let state = { date: today, count: 0 };
    if (raw) {
      const parsed = JSON.parse(raw);
      if (parsed?.date === today) state = parsed;
    }
    state.count += 1;
    localStorage.setItem(BYOK_COUNTER_KEY, JSON.stringify(state));
    return state.count;
  } catch {
    return 1;
  }
}

export const BYOK_TUNING = { BYOK_NUDGE_AT, BYOK_DAILY_SOFT_LIMIT };

function mockGenerate(req: GenerateNamesRequest): GenerateNamesResponse {
  const before = readUsage();
  const cooldownLeft = before.lastGeneratedAt
    ? Math.max(0, TIER_LIMITS.engagedCooldownMs - (Date.now() - before.lastGeneratedAt))
    : 0;
  if (tierFor(before.count) === "engaged" && cooldownLeft > 0) {
    return assembleMock(req, before, cooldownLeft, true);
  }

  const state = incrementUsage();
  return assembleMock(req, state, 0, false);
}

function assembleMock(
  req: GenerateNamesRequest,
  state: UsageState,
  cooldownMs: number,
  cachedReplay: boolean,
): GenerateNamesResponse {
  const tier = tierFor(state.count);
  const tlds = TLD_BUCKETS[req.tldPreference ?? "any"];
  const tokens = extractTokens(req.description);
  const bases = buildBases(tokens, req.styles ?? []);
  const limit = TIER_LIMITS.fullResultCount;
  const all: GeneratedName[] = [];

  for (const base of bases) {
    for (const tld of tlds) {
      if (all.length >= limit) break;
      const domain = `${base}.${tld}`;
      all.push({
        domain,
        available: pseudoAvailable(domain),
        tld,
        base,
        rationale: rationaleFor(base, req.styles ?? []),
      });
    }
    if (all.length >= limit) break;
  }

  const visible = tier === "engaged" ? TIER_LIMITS.engagedVisibleCount : all.length;
  const results = all.map((r, i) => (i < visible ? r : { ...r, locked: true }));

  return {
    results,
    description: req.description,
    listId: `mock-${state.count}`,
    tier,
    generationsUsed: state.count,
    generationsLimit:
      tier === "curious"
        ? TIER_LIMITS.curiousLimit
        : tier === "trying"
        ? TIER_LIMITS.tryingLimit
        : TIER_LIMITS.tryingLimit,
    signupUrl: withVisitorId(`${PORTAL_URL}/signup`),
    cooldownMs: cachedReplay ? cooldownMs : tier === "engaged" ? TIER_LIMITS.engagedCooldownMs : 0,
    mode: "hosted",
  };
}

const STOPWORDS = new Set([
  "a", "an", "and", "the", "for", "to", "of", "in", "on", "with", "that",
  "this", "is", "it", "we", "our", "i", "my", "be", "by", "as", "at", "or",
  "from", "but", "are", "was", "were", "will", "would", "can", "could",
  "should", "have", "has", "had", "you", "your", "they", "their", "them",
  "app", "platform", "service", "company", "business", "startup",
]);

function extractTokens(text: string): string[] {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9 ]/g, " ")
    .split(/\s+/)
    .filter((t) => t.length >= 3 && !STOPWORDS.has(t))
    .slice(0, 8);
}

function buildBases(tokens: string[], styles: NameStyle[]): string[] {
  const bases = new Set<string>();
  const playful = styles.includes("playful");
  const short = styles.includes("short");
  const compound = styles.includes("compound");
  const wordplay = styles.includes("wordplay");

  for (const t of tokens) {
    bases.add(t);
    if (short) bases.add(t.slice(0, Math.max(4, Math.floor(t.length * 0.7))));
    if (playful) {
      bases.add(`${t}ly`);
      bases.add(`${t}oo`);
      bases.add(`get${t}`);
    }
    if (wordplay) {
      bases.add(t.replace(/s$/, "z"));
      bases.add(`${t}ify`);
    }
  }

  if (compound || tokens.length >= 2) {
    for (let i = 0; i < tokens.length; i++) {
      for (let j = 0; j < tokens.length; j++) {
        if (i === j) continue;
        bases.add(`${tokens[i]}${tokens[j]}`);
      }
    }
  }

  const PREFIXES = ["use", "try", "join", "go"];
  const SUFFIXES = ["hq", "labs", "io", "stack", "kit", "hub"];
  for (const t of tokens.slice(0, 3)) {
    for (const p of PREFIXES) bases.add(`${p}${t}`);
    for (const s of SUFFIXES) bases.add(`${t}${s}`);
  }

  return Array.from(bases).filter((b) => b.length >= 3 && b.length <= 20);
}

function pseudoAvailable(domain: string): boolean | null {
  let h = 0;
  for (let i = 0; i < domain.length; i++) h = (h * 31 + domain.charCodeAt(i)) | 0;
  const v = Math.abs(h) % 10;
  if (v < 6) return true;
  if (v < 9) return false;
  return null;
}

function rationaleFor(base: string, styles: NameStyle[]): string {
  if (styles.includes("playful")) return `Playful spin on "${base}"`;
  if (styles.includes("short")) return `Short and memorable`;
  if (base.length <= 6) return `Short, brandable`;
  return `Direct and descriptive`;
}
